libraries/AP_WayBack/AP_WayBack.h

```diff
@@ -0,0 +1,257 @@
+ unnecessary data - redundant points and loops, so it always knows 
+ the shortest way to home via visited points.
+ 
+ Copyright night_ghost@ykoctpa.ru 2017
+  
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#pragma once
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_AHRS/AP_AHRS.h>
+
+#define NUM_TRACK_POINTS    4000     // 4000 points by 8 bytes = 32K of RAM. Feel free to change size
+#define SIMPLIFY_MOVE_BACK     2  // move N points back
+#define MIN_SIMPLIFY_POINTS    8 // this number should be adjusted upto value which not cause processing time to exceed 1s at full load
+#define RAW_POINTS             3
+
+#ifdef WAYBACK_DEBUG
+  #define DBG_PRINTLN(x)     { hal.uartA->print("#"); hal.uartA->print(x); hal.uartA->println();  }
+  #define DBG_PRINT(x)       { hal.uartA->print("#"); hal.uartA->print(x); hal.uartA->print(" "); }
+  #define DBG_PRINTVARLN(x)  { hal.uartA->print("#" #x); hal.uartA->print(": "); hal.uartA->println(x); }
+  #define DBG_PRINTVAR(x)    { hal.uartA->print("#" #x); hal.uartA->print(": "); hal.uartA->print(x); hal.uartA->print(" ");  }
+  #define DBG_PRINTF(x,...)  { hal.uartA->printf("#" x,## __VA_ARGS__); }
+#else
+     #define DBG_PRINTLN(x)     {}
+     #define DBG_PRINT(x)       {}
+     #define DBG_PRINTVARLN(x)  {}
+     #define DBG_PRINTVAR(x)    {}
+     #define DBG_PRINTF(x,...)  {}
+#endif
+
+#define LOOPS_MOVE_BACK 12 // new segments not checked for loops
+
+#define EPS_DISTANCE 1 // in meters, by default
+
+#define SHORTCUT_FACTOR 2 // на сколько чувствительность "близости" точек больше чем Эпсилон
+
+#if FRAME_CONFIG == MULTICOPTER_FRAME
+
+// copter's coordinates in meters
+#define TRACK_EPS (_params.eps) // meters 
+
+#else
+
+// all another coordinates are from GPS so 1 gradus is ~110KM 
+#define TRACK_EPS (_params.eps * 110000) //  meters in gradus of coords 
+
+#endif
+
+template<typename T> void zeroIt(T& value)
+{
+    memset(&value,0,sizeof(value));
+}
+
+/* _end is set in the linker command file */
+extern uint8_t * _end;
+
+
+class AP_WayBack
+{
+public:
+
+    enum POINT_STATUS {
+        POINT_EMPTY=0,
+        POINT_OK,
+        POINT_LOOP
+    };
+
+#pragma pack(push, 1)
+    typedef struct POINT {
+        float x;
+        float y;
+//        enum POINT_STATUS status; // this field should be changed to NAN in X coord as boolean flag
```

Comment:
We have already discussed this issue before. The world does not have a roof so the return altitude doea as it was set.